src/core/lwt.ml

```diff
@@ -22,1307 +22,2777 @@
+   corresponding match case [Proxy _] is impossible.
+
+   The external promise type, ['a t], and the external resolver type, ['a u],
+   are not GADTs. Furthermore, they are, respectively, covariant and
+   contravariant in ['a], while the internal promise type is invariant in
+   ['a]. For these reasons, there are nasty casts between ['a t], ['a u], and
+   the internal promise type. The implementation is, of course, written in
+   terms of the internal type.
+
+   Casting from an ['a t] to an internal promise produces a reference for
+   which the state is "unknown": this is simulated with a helper GADT, which
+   encodes existential types. There are several similar casts, which are used
+   to document possible state changes between the time a promise is created,
+   and the later time it is used in a callback. You can see these casts in
+   action in [Lwt.bind]. The cast syntax is pretty light, and, besides being
+   commented in [bind], all such casts are documented in modules [Public_types]
+   and [Basic_helpers].
+
+   There is an abstract type [in_completion_loop], which is actually just
+   [unit], that is passed around to help ensure that certain functions can only
+   be called during the completion loop. Those functions can't be called without
+   a value of type [in_completion_loop], and the only way to obtain such a value
+   is to enter the loop ([run_in_completion_loop]). This mechanism is described
+   at [Completion_loop.complete].
 
-type a_thread = (module A_thread)
-    (* a_thread = exists 'a. 'a t *)
 
-(* Pack a thread into a A_thread module. *)
-let pack_thread (type x) t =
-  let module M = struct type a = x let thread = t end in
-  (module M : A_thread)
+   If you've made it this far, you are an Lwt expert! Rejoice! *)
 
-(* A type + a list of threads of this type. *)
-module type A_threads = sig
-  type a
-  val threads : a t list
+
+
+(* Some sequence-associated storage types
+
+   Sequence-associated storage is defined and documented later, in module
+   [Sequence_associated_storage]. However, the following types are mentioned in
+   the definition of [promise], so they must be defined here first. *)
+module Storage_map =
+  Map.Make
+    (struct
+      type t = int
+      let compare = compare
+    end)
+type storage = (unit -> unit) Storage_map.t
+
+
+
+module Main_internal_types =
+struct
+  (* Phantom types for use with types [promise] and [state]. These are never
+     constructed; the purpose of the constructors is to prove to the type
+     checker that these types are distinct from each other. Warning 37, "unused
+     constructor," therefore has to be temporarily suppressed. *)
+
+  [@@@ocaml.warning "-37"]
+
+  type underlying = Underlying_and_this_constructor_is_not_used
+  type proxy = Proxy_and_this_constructor_is_not_used
+
+  type completed = Completed_and_this_constructor_is_not_used
+  type pending = Pending_and_this_constructor_is_not_used
```

Comment:
Very nice, I've done so.

This is actually enforcing even what *prevented* me from making the promise record type private – that the record would become immutable even inside the module `Main_internal_types`. `private` in a module is perfect for phantom types, though.